cmd/segment-reaper/observer_test.go

```diff
@@ -0,0 +1,501 @@
+					obj := obsvr.objects[bucketName][objPath]
+					assert.Equal(t, numSegments, int(obj.expectedNumberOfSegments), "Object.expectedNumSegments")
+					assert.True(t, obj.hasLastSegment, "Object.hasLastSegment")
+					assert.False(t, obj.skip, "Object.skip")
+				}
+			}
+		}
+
+		// Assert observer global stats
+		assert.Zero(t, obsvr.inlineSegments, "inlineSegments")
+		assert.Zero(t, obsvr.lastInlineSegments, "lastInlineSegments")
+		assert.Equal(t, numSegments, obsvr.remoteSegments, "remoteSegments")
+	})
+
+	t.Run("objects with at least one has more than 65 segments", func(t *testing.T) {
+		ctx := testcontext.New(t)
+		defer ctx.Cleanup()
+
+		var (
+			testdata = generateTestdataObjects(t, ctx.Context, false, true)
+			obsvr    = observer{
+				objects: make(bucketsObjects),
+			}
+		)
+
+		for _, objSeg := range testdata.objSegments {
+			err := obsvr.processSegment(ctx.Context, objSeg.path, objSeg.pointer)
+			require.NoError(t, err)
+		}
+
+		// Assert observer internal state
+		assert.Equal(t, testdata.projectID.String(), obsvr.lastProjectID, "lastProjectID")
+		if assert.Equal(t, len(testdata.expectedObjects), len(obsvr.objects), "objects number") {
+			for bucket, bucketObjs := range obsvr.objects {
+				expBucketObjs, ok := testdata.expectedObjects[bucket]
+				if !ok {
+					t.Errorf("bucket '%s' shouldn't exist in objects map", bucket)
+					continue
+				}
+
+				if !assert.Equalf(t, len(expBucketObjs), len(bucketObjs), "objects per bucket (%s) number", bucket) {
+					continue
+				}
+
+				for expPath, expObj := range expBucketObjs {
+					if !assert.Contains(t, bucketObjs, expPath, "path in bucket objects map") {
+						continue
+					}
+
+					obj := bucketObjs[expPath]
+					assert.Equal(t, expObj.expectedNumberOfSegments, obj.expectedNumberOfSegments, "Object.expectedNumSegments")
+					assert.Equal(t, expObj.hasLastSegment, obj.hasLastSegment, "Object.hasLastSegment")
+
+					assert.Equal(t, expObj.skip, obj.skip, "Object.skip")
+					if !expObj.skip {
+						assert.Equal(t, expObj.segments, obj.segments, "Object.segments")
+					}
+				}
+			}
+		}
+
+		// Assert observer global stats
+		assert.Equal(t, testdata.expectedInlineSegments, obsvr.inlineSegments, "inlineSegments")
+		assert.Equal(t, testdata.expectedInlineSegments, obsvr.lastInlineSegments, "lastInlineSegments")
+		assert.Equal(t, testdata.expectedRemoteSegments, obsvr.remoteSegments, "remoteSegments")
+	})
+}
+
+// segmentRef is an object segment reference to be used for simulating calls to
+// observer.processSegment
+type segmentRef struct {
+	path    metainfo.ScopedPath
+	pointer *pb.Pointer
+}
+
+// createNewObjectSegments creates a list of segment references which belongs to
+// a same object.
+//
+// If inline is true the last segment will be of INLINE type.
+//
+// If withNumSegments is true the last segment pointer will have the
+// NumberOfSegments set.
+//
+// It returns the object path and the list of object segment references.
+//nolint:golint
```

Comment:
Why are we disabling the linter here? (and on another place too)

Reply:
Linter doesn't like functions that have context parameter and it isn't the first. In test, I prefer to have as a first parameter the `testing.T` rather than `context.Context`, but if we prefer the opposite in order of not adding the `//nolint` comment let me know and I change them.

Reply:
I'd prefer to keep the linter enabled. This test code is really complex, so let's have the linter helping us.

Reply:
ok no problem, only mention that the linter is only disabled for the following line, not further than that.